cmd/snap-update-ns/change.go

```diff
@@ -57,28 +58,180 @@ func (c Change) String() string {
+			pokeHoles = false
+			goto retry
+		}
+	}
+	return changes, err
+}
+
+func (c *Change) ensureTarget() ([]*Change, error) {
+	var changes []*Change
+
+	kind, _ := c.Entry.OptStr("x-snapd.kind")
+	path := c.Entry.Dir
+
+	// We use lstat to ensure that we don't follow a symlink in case one was
+	// set up by the snap. Note that at the time this is run, all the snap's
+	// processes are frozen but if the path is a directory controlled by the
+	// user (typically in /home) then we may still race with user processes
+	// that change it.
+	fi, err := osLstat(path)
+
+	if err == nil {
+		// If the element already exists we just need to ensure it is of
+		// the correct type. The desired type depends on the kind of entry
+		// we are working with.
+		switch kind {
+		case "":
+			if !fi.Mode().IsDir() {
+				err = fmt.Errorf("cannot use %q for mounting, not a directory", path)
+			}
+		case "file":
+			if !fi.Mode().IsRegular() {
+				err = fmt.Errorf("cannot use %q for mounting, not a regular file", path)
+			}
+		case "symlink":
+			// When we want to create a symlink we just need the empty
+			// space so anything that is in the way is a problem.
+			err = fmt.Errorf("cannot create symlink in %q, existing file in the way", path)
+		}
+	} else if os.IsNotExist(err) {
+		// No such file or directory, create it
+		if kind == "symlink" {
+			// For symlinks the path should refer to the parent directory.
+			// We set it here so that it gets picked up by the error message
+			// below, in case things fail.
+			path = filepath.Dir(c.Entry.Dir)
 		}
-		// If this is a bind mount then create the source directory as well.
-		// This allows snaps to share a subset of their data easily.
-		flags, _ := mount.OptsToCommonFlags(c.Entry.Options)
-		if flags&syscall.MS_BIND != 0 {
-			if err := ensureMountPoint(c.Entry.Name, mode, uid, gid); err != nil {
-				return nil, err
+		changes, err = c.createInodes(path, kind, true)
+	} else {
+		// If we cannot inspect the element let's just bail out.
+		err = fmt.Errorf("cannot inspect %q: %v", path, err)
+	}
+	return changes, err
+}
+
+func (c *Change) ensureSource() error {
+	// We only have to do ensure bind mount source exists.
+	// This also rules out symlinks.
+	flags, _ := mount.OptsToCommonFlags(c.Entry.Options)
+	if flags&syscall.MS_BIND == 0 {
+		return nil
+	}
+
+	kind, _ := c.Entry.OptStr("x-snapd.kind")
+	path := c.Entry.Name
+	fi, err := osLstat(path)
+
+	if err == nil {
+		// If the element already exists we just need to ensure it is of
+		// the correct type. The desired type depends on the kind of entry
+		// we are working with.
+		switch kind {
+		case "":
+			if !fi.Mode().IsDir() {
+				err = fmt.Errorf("cannot use %q for mounting, not a directory", path)
+			}
+		case "file":
+			if !fi.Mode().IsRegular() {
+				err = fmt.Errorf("cannot use %q for mounting, not a regular file", path)
 			}
 		}
+	} else if os.IsNotExist(err) {
+		_, err = c.createInodes(path, kind, false)
```

Comment:
Why are we creating the source location?  Shouldn't we just force it to exist?

Reply:
(replied during hangout) - to allow sharing directories / files before the owner of the files started and created them.